src/miner_port_services_sup.erl

```diff
@@ -0,0 +1,99 @@
+            SupFlags = #{
+                strategy => rest_for_one,
+                intensity => 0,
+                period => 1
+            },
+
+            GatewayTcpPort = application:get_env(miner, gateway_api_port, 4468),
+
+            KeyPair =
+                case application:get_env(blockchain, key, undefined) of
+                    undefined ->
+                        BaseDir = application:get_env(blockchain, base_dir, "data"),
+                        GatewayKey = filename:absname(filename:join([BaseDir, "miner", "gateway_swarm_key"])),
+                        ok = filelib:ensure_dir(GatewayKey),
+                        case filelib:is_file(GatewayKey) of
+                            true ->
+                                GatewayKey;
+                            false ->
+                                SwarmKey = filename:join([BaseDir, "miner", "swarm_key"]),
+                                case filelib:is_file(SwarmKey) of
+                                    true ->
+                                        {ok, KeyMap} = libp2p_crypto:load_keys(SwarmKey),
+                                        {ok, GatewayKeyMap} = miner_keys:libp2p_to_gateway_key(KeyMap),
+                                        ok = libp2p_crypto:save_keys(GatewayKeyMap, GatewayKey),
+                                        GatewayKey;
+                                    false ->
+                                        KeyMap = libp2p_crypto:generate_keys(ecc_compact),
+                                        ok = libp2p_crypto:save_keys(KeyMap, SwarmKey),
+                                        {ok, GatewayKeyMap} = miner_keys:libp2p_to_gateway_key(KeyMap),
+                                        ok = libp2p_crypto:save_keys(GatewayKeyMap, GatewayKey),
+                                        GatewayKey
+                                end
+                        end;
+                    {ecc, Keypair0} ->
+                        case io_lib:char_list(Keypair0) of
+                            true -> Keypair0;
+                            false -> miner_keys:key_proplist_to_uri(Keypair0)
+                        end
+                end,
+
+            GatewayECCWorkerOpts = [
+                {transport, application:get_env(miner, gateway_transport, tcp)},
+                {host, application:get_env(miner, gateway_host, "localhost")},
+                {port, GatewayTcpPort}
+            ],
+
+            GatewayPortOpts = [{keypair, KeyPair}] ++ GatewayECCWorkerOpts,
+
+            MuxOpts = [
+                {host_port, application:get_env(miner, mux_host_port, 1680)},
+                {client_ports, application:get_env(miner, mux_client_ports, [1681, 1682])}
+            ],
+
+            ChildSpecs =
+                [
+                    ?WORKER(miner_gateway_port, [GatewayPortOpts]),
+                    ?WORKER(miner_gateway_ecc_worker, [GatewayECCWorkerOpts]),
+                    ?WORKER(miner_mux_port, [MuxOpts])
```

Comment:
i put the mux last so the gateway was up and running before the mux was able to send it any packets, the alternative being if the mux were first it could be routing packets to the gateway that weren't being handled yet while it was still starting.
if the mux dies it should get restarted by the supervisor and in the meantime the other two just aren't receiving any packets so they're just waiting as if the mux were running all along and not receiving packets